book/src/proposals/validator-timestamp-oracle.md

```diff
@@ -0,0 +1,118 @@
+# Validator Timestamp Oracle
+
+Third-party users of Solana sometimes need to know the real-world time a block
+was produced, generally to meet compliance requirements for external auditors or
+law enforcement. This proposal describes a validator timestamp oracle that
+would allow a Solana cluster to satisfy this need.
+
+The general outline of the proposed implementation is as follows:
+
+- At regular intervals, each validator records its observed time for a known slot
+  on-chain (via a Timestamp added to a slot Vote)
+- A client can request a block time using the `getBlockTime` RPC method. When a
+  client requests a timestamp for block N:
+
+  1. A validator determines a "cluster" timestamp for a recent slot before block N
+  by observing all the Timestamp instructions recorded on the ledger that
+  reference that slot, and determining the stake-weighted mean timestamp.
+
+  2. This recent mean timestamp is then used to calculate the timestamp of
+  block N using the cluster's established slot duration
+
+Requirements:
+- Any validator replaying the ledger in the future must come up with the same
+  time for every block since genesis
+- Estimated block times should not drift more than an hour or so before resolving
+  to real-world (oracle) data
+- The block times are not controlled by a single centralized oracle, but
+  ideally based on a function that uses inputs from all validators
+- Each validator must maintain a timestamp oracle
+
+## Recording Time
+
+At regular intervals, each validator records its observed time by submitting a
+VoteWithTimestamp instruction to the cluster as it is voting on a particular
```

Comment:
Thinking more on this, instead of `VoteWithTimestamp`, I'd rather see an `Option<Timestamp>` where `Timestamp` is a `u64`, seconds since the Unix Epoch.  That costs us 1 byte per Vote transaction (when `None`), but avoids having to duplicate the instruction. We may want to add more fields in the future, and it'd be a source of bugs if we had to add that field in 2 places.

Reply:
Ah, nice, agreed. Updating...